### DISCUSSION TITLE
How does composite index work in SQLModel? Query part

### DISCUSSION THREAD

**Discussion by tepelbaum** [fastapi/sqlmodel#1134] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the SQLModel documentation, with the integrated search.
- [X] I already searched in Google "How to X in SQLModel" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to SQLModel but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to SQLModel but to [SQLAlchemy](https://github.com/sqlalchemy/sqlalchemy).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
"""
The class definition
"""
from sqlmodel import Field
from sqlmodel import Index
from sqlmodel import SQLModel



class Foo(SQLModel, table=True):  # type: ignore
    """
    Exposure db version
    """
    __tablename__ = 'foo'
    id: int | None = Field(default=None, primary_key=True)
    bar_1: str
    bar_2: str
    bar_3: str

    __table_args__ = (
        Index(
            'foo_filter_id',
            'bar_1',
            'bar_2',
            'bar_3'
        ),
    )

"""
Notebook imports
"""
from ecodev_core.db_connection import DB_URL, DB
from sqlmodel import create_engine
engine= create_engine(DB_URL,echo=True)
import psycopg2 as psy
from psycopg2 import extras as psy_extras
from app.db_model import Foo
from sqlmodel import SQLModel, select, Session
conf = {
        'host': DB.db_host,
        'database': DB.db_name,
        'user': DB.db_username,
        'password': DB.db_password,
        'port': DB.db_port,
    }
def db_cursor():
    """
        Get database connection adn cursor based on config file
    """
    connection = psy.connect(**conf)
    return connection, connection.cursor(cursor_factory=psy_extras.RealDictCursor)


"""
psycopg2 direct call, working (meaning using the index)
"""

%%time
db_con, db_cur = db_cursor()
db_cur.execute(f"explain SELECT * from foo where bar_1 = 'a' and bar_2='g' and bar_3 = 's'")
db_cur.fetchall()


"""
SQLModel not working, meaning not using the index
"""
%%time
with Session(engine) as session:
    toto = session.exec(select(Foo.bar_1,Foo.bar_2,Foo.bar_3).where(
        Foo.bar_1 == 'a',
        Foo.bar_2 ==  'g',
        Foo.bar_3 == 's')).all()
```

### Description

Hi. 

I successfully created a composite index (see code snippet), but can't manage to forge a  query that actually uses it.

Some screenshots 
![Capture d’écran du 2024-10-11 16-17-38](https://github.com/user-attachments/assets/a285e742-d6c2-4e8a-ad8c-a1a4e475bf89)

index indeed generated via sqlmodel
![image](https://github.com/user-attachments/assets/151f0073-de39-461f-9989-6d988e1f38b1)

composite index working in plain sql
![Capture d’écran du 2024-10-11 16-18-18](https://github.com/user-attachments/assets/cce55b34-36e1-4f0f-b7bb-ec66f235430c)

composite index working in psycopg2
![Capture d’écran du 2024-10-11 16-25-57](https://github.com/user-attachments/assets/da180371-007c-4795-9168-a04fbc39493a)

composite index not working in SQLModel
![Capture d’écran du 2024-10-11 16-27-12](https://github.com/user-attachments/assets/06f777d0-2984-4a7d-99fd-d03f40fede2d)

### Operating System

Linux

### SQLModel Version

0.0.22

### Python Version

3.11

**Comment by YuriiMotov** [comment, marked as answer]
Why do you think it doesn't work?
You didn't use `explain` statement in case of SQLModel query. If you try you will see it works:

```py
from sqlmodel import Field, Index, Session, SQLModel, col, create_engine, select, text

class Foo(SQLModel, table=True):  # type: ignore
    """
    Exposure db version
    """
    __tablename__ = "foo"
    id: int | None = Field(default=None, primary_key=True)
    bar_1: str
    bar_2: str
    bar_3: str

    __table_args__ = (Index("foo_filter_id", "bar_1", "bar_2", "bar_3"),)


engine = create_engine(
    "postgresql://user:mysecretpassword@localhost:5432/some_db", echo=False
)
SQLModel.metadata.create_all(engine)


with Session(engine) as session:
    st = (
        select(Foo.bar_1, Foo.bar_2, Foo.bar_3)
        .where(col(Foo.bar_1) == "a")
        .where(Foo.bar_2 == "g")
        .where(Foo.bar_3 == "s")
    )
    st_compiled = st.compile(engine, compile_kwargs={"literal_binds": True})
    res = session.exec(text(f"EXPLAIN {st_compiled}")).all()
    print(res)

```
Output:
```
[
    ('Index Only Scan using foo_filter_id on foo  (cost=0.15..8.17 rows=1 width=96)',),
    ("  Index Cond: ((bar_1 = 'a'::text) AND (bar_2 = 'g'::text) AND (bar_3 = 's'::text))",)
]
```
